gyronimo/dynamics/cartesian_boris.cc

```diff
@@ -128,51 +86,36 @@ IR3 cartesian_boris::get_velocity(const state& s) const {
 
 //! Returns the kinetic energy of the state, normalized to `Uref`.
 double cartesian_boris::energy_kinetic(const state& s) const {
-  return s[3] * s[3] + s[4] * s[4] + s[5] * s[5];
+  IR3 v = this->get_velocity(s);
+  return inner_product(v, v);
 }
 
 //! Returns the parallel energy of the state, normalized to `Uref`.
 double cartesian_boris::energy_parallel(const state& s, double& time) const {
-  IR3 x = {s[0], s[1], s[2]};
-  IR3 v = {s[3], s[4], s[5]};
-  double Btime = time * iBfield_time_factor_;
-  IR3 b = magnetic_field_->contravariant_versor(x, Btime);
-  double vpar = inner_product(v, b);
-  return vpar * vpar;
+  IR3 x = this->get_position(s), v = this->get_velocity(s);
+  IR3 b = magnetic_field_->contravariant_versor(x, time * iB_time_factor_);
+  double v_parallel = inner_product(v, b);
+  return v_parallel * v_parallel;
 }
 
 //! Returns the perpendicular energy of the state, normalized to `Uref`.
 double cartesian_boris::energy_perpendicular(
     const state& s, double& time) const {
-  IR3 x = {s[0], s[1], s[2]};
-  IR3 v = {s[3], s[4], s[5]};
-  double Btime = time * iBfield_time_factor_;
-  IR3 b = magnetic_field_->contravariant_versor(x, Btime);
-  IR3 vperp = cross_product(v, b);
-  return inner_product(vperp, vperp);
+  IR3 x = this->get_position(s), v = this->get_velocity(s);
+  IR3 b = magnetic_field_->contravariant_versor(x, time * iB_time_factor_);
+  IR3 v_perpendicular = cross_product(v, b);
+  return inner_product(v_perpendicular, v_perpendicular);
 }
 
-cartesian_boris::state cartesian_boris::generate_initial_state(
-    const IR3& pos, const IR3& vel, const double& time,
-    const double& dt) const {
-  // state s0 = {
-  //     pos[IR3::u], pos[IR3::v], pos[IR3::w],
-  //     vel[IR3::u], vel[IR3::v], vel[IR3::w]
-  // };
-  // state s1 = do_step(s0, time, -0.5*dt);
-
-  // return {s0[0], s0[1], s0[2], s1[3], s1[4], s1[5]};
-
-  lorentz lo(Lref_, Vref_, qom_, electric_field_, magnetic_field_);
-  odeint_adapter<gyronimo::lorentz> sys(&lo);
+//! Returns a state with the velocity integrated backwards by a half time step.
+cartesian_boris::state cartesian_boris::half_back_step(
+    const IR3& x, const IR3& v, const double& time, const double& dt) const {
+  lorentz lo(Lref_, Vref_, qom_, magnetic_field_, electric_field_);
+  auto ls = lo.generate_state(x, v);
   boost::numeric::odeint::runge_kutta4<gyronimo::lorentz::state> rk4;
-  lorentz::state inout = {pos[IR3::u], pos[IR3::v], pos[IR3::w],
-                          vel[IR3::u], vel[IR3::v], vel[IR3::w]};
-  rk4.do_step(sys, inout, time, -0.5 * dt);
-  IR3 vmh = {inout[3], inout[4], inout[5]};
-
-  return {pos[IR3::u], pos[IR3::v], pos[IR3::w],
-          vmh[IR3::u], vmh[IR3::v], vmh[IR3::w]};
+  rk4.do_step(odeint_adapter(&lo), ls, time, -0.5 * dt);
+  IR3 x_half_back = lo.get_position(ls), v_half_back = lo.get_velocity(ls);
```

Comment:
`x_half_back` is an unused variable since it is not returned.

Reply:
Aye, will be fixed.